File: tests/test-minecraft-osp/src/main/java/pize/tests/voxelgame/main/text/Component.java
package pize.tests.voxelgame.main.text;

import pize.graphics.util.color.Color;

import java.util.ArrayList;
import java.util.List;

public class Component{
    
    private final List<Component> components;
    private final TextStyle style;
    private final Color color;
    
    public Component(TextStyle style, Color color){
        this.components = new ArrayList<>();
        this.style = style.copy();
        this.color = color.copy();
    }
    
    public Component(){
        this.components = new ArrayList<>();
        this.style = new TextStyle();
        this.color = new Color();
    }
    
    
    
    public Component component(Component component){
        components.add(component);
        return this;
    }
    
    
    public Component style(StyleFormatting styleFormatting){
        style.enable(styleFormatting);
        return this;
    }
    
    public Component color(TextColor textColor){
        this.color.set(textColor.color);
        return this;
    }
    
    public Component color(double r, double g, double b){
        this.color.set(r, g, b, 1.0);
        return this;
    }
    
    public Component reset(){
        style.reset();
        color.reset();
        return this;
    }
    
    public Component text(String text){
        final ComponentText component = new ComponentText(style, color, text);
        components.add(component);
        return this;
    }
    
    public Component text(Object objectToString){
        return text(String.valueOf(objectToString));
    }
    
    public Component translate(String translateKey, Component... arguments){
        final ComponentTranslate component = new ComponentTranslate(style, color, translateKey, arguments);
        components.add(component);
        return this;
    }
    
    public Component formattedText(String text){
        text(text);
        return this;
    }
    
    
    @Override
    public String toString(){
        final StringBuilder builder = new StringBuilder();
        for(Component component: components)
            builder.append(component.toString());
    
        return builder.toString();
    }
    
    
    public List<ComponentText> toFlatList(){
        final List<ComponentText> list = new ArrayList<>();
        this.addToFlatList(list);
        return list;
    }
    
    private void addToFlatList(List<ComponentText> list){
        for(Component component: components){
            if(component instanceof ComponentText componentText)
                list.add(componentText);
            
            component.addToFlatList(list);
        }
    }
    
    
    public void addComponent(Component component){
        components.add(component);
    }
    
    public TextStyle getStyle(){
        return style;
    }
    
    public Color getColor(){
        return color;
    }
    
}

File: tests/test-minecraft-osp/src/main/kotlin/pize/tests/voxelgame/main/text/Component.kt
package pize.tests.voxelgame.main.text

import pize.graphics.util.color.Color

open class Component {
    private val components: MutableList<Component>
    val style: TextStyle?
    val color: Color

    constructor(style: TextStyle?, color: Color) {
        components = ArrayList()
        this.style = style!!.copy()
        this.color = color.copy()
    }

    constructor() {
        components = ArrayList()
        style = TextStyle()
        color = Color()
    }

    fun component(component: Component): Component {
        components.add(component)
        return this
    }

    fun style(styleFormatting: StyleFormatting): Component {
        style!!.enable(styleFormatting)
        return this
    }

    fun color(textColor: TextColor): Component {
        color.set(textColor.color)
        return this
    }

    fun color(r: Double, g: Double, b: Double): Component {
        color[r, g, b] = 1.0
        return this
    }

    fun reset(): Component {
        style!!.reset()
        color.reset()
        return this
    }

    fun text(text: String?): Component {
        val component = ComponentText(style, color, text)
        components.add(component)
        return this
    }

    fun text(objectToString: Any?): Component {
        return text(objectToString.toString())
    }

    fun translate(translateKey: String?, vararg arguments: Component?): Component {
        val component = ComponentTranslate(style, color, translateKey, *arguments)
        components.add(component)
        return this
    }

    fun formattedText(text: String?): Component {
        text(text)
        return this
    }

    override fun toString(): String {
        val builder = StringBuilder()
        for (component in components) builder.append(component.toString())
        return builder.toString()
    }

    fun toFlatList(): MutableList<ComponentText> {
        val list: MutableList<ComponentText> = ArrayList()
        addToFlatList(list)
        return list
    }

    private fun addToFlatList(list: MutableList<ComponentText>) {
        for (component in components) {
            if (component is ComponentText) list.add(component)
            component.addToFlatList(list)
        }
    }

    fun addComponent(component: Component) {
        components.add(component)
    }
}
